Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78523
* #78522
* #78677
* #78536
* __->__ #78090

This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

Commit:
Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"




[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"




[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"




[ghstack-poisoned]

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -44,6 +45,41 @@ def _is_tensor_constructor(func: OpOverload):
+
+# if someone invokes `FakeTensorMode` multiple times, we should
+# only clear the fake tensor converter when all modes are done
+active_fake_tensor_modes = 0
```
There's an argument for using a stack here... happy to do either

ezyang replied:
This doesn't look compositional. But I know that you were also wondering what it meant to be compositional so let me try to explain.

The primary characteristic of compositionality is that you can reason about various transforms independently and by doing so understand with the whole system does. In other words if there is a fake tensor mode you can understand what it does and it doesn't have any spooky at a distance interactions with further fake tensor modes elsewhere on the stack. The model of compositionality for modes is that any given mode will transform function calls to other function calls, which then get handled by the next mode on the stack. That is to say I could eliminate any given mode simply by rewriting my program to look like the result of having applied that transformation.

With this in mind let's think about what is going on with the code here. You do have action at a distance: multiple instances of the mode share the same converter. You can't just eliminate one of the modes without changing the behavior of the other mode. What I have suggested you do instead is to maintain a separate converter per recursive mode. This way there is no interaction between the modes. I don't really know under what circumstances a recursive mode would be useful here, but I don't really need to know about the use case because the principle of composition tells me what the behavior should be.

Does that help?

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. 



[ghstack-poisoned]

albanD commented:
> Disables the debug invariant checking in VariableType that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage. These assertions limit what you can do with torch_dispatch_mode, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

I'm not sorry we want to do that? These assertions are important and the view behavior must be followed. While I agree that they are annoying, user code might get silently wrong if you do not respect these.

eellison commented:
@albanD 
> I'm not sorry we want to do that? These assertions are important and the view behavior must be followed. While I agree that they are annoying, user code might get silently wrong if you do not respect these.

sorry it just disables them when a torch dispatch mode is set

ezyang commented:
I'm a bit surprised that you needed the asserts to be disabled for this patch; it feels like part of the point of conversion caching (ugh, I hate this name, it implies that you can drop the caching and the result will be sound but that's not the case here) is to make sure the storage relationships are setup appropriately so the asserts don't fail.

eellison commented:
> I'm a bit surprised

In this call
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
```

`y` won't share a storage with `x`, because `y` will have a `meta` storage and `x` won't. So it will fail `ENFORCE_SAME_TENSOR_STORAGE`.

If you wanted to do something simple like hold onto every tensor that gets run through the backward, and save it with its corresponding op usages, you would fail the `AT_ASSERT(${tensor_name}.use_count() <= 1, "function: ${fn_name}");` call.

albanD commented:
Isn't a simple fix to create the Tensor inside your context?
We might consider an option to disable them. But that should definitely not be the default. In your example, any inplace op will be just wrong.
Note that if the idea is that you are re-implementing the view semantic but don't use storage for that (which trips this test). Then I think we are happy to add new API to fix that. Most like you want to do like functionalization and have a fake storage that you re-use to show that your Tensors share data. That fake storage is also a good place for you to add your logic that reproduces the view behavior as it will be shared by all the Tensors that are supposed to be views of each other.

eellison commented:
> Isn't a simple fix to create the Tensor inside your context?
This doesn't mimic the existing `FakeTensor` mode usage, and I think would be cumbersome. 

> In your example, any inplace op will be just wrong 
Something like `add_` only changes the values of the tensors, which as far as fake tensors go is irrelevant, so I wouldn't consider it to be wrong. for `resize_` etc, we would probably want to throw if the inputs are not FakeTensors already. I think this would be an unlikely case, as you're unlikely to resize_ on your module parameters or inputs. 

> Most like you want to do like functionalization and have a fake storage that you re-use to show that your Tensors share data 

Link me ? I don't want to duplicate the conversion to meta storage that already exists so we should get alignment with @ezyang if this is the approach we want to go down.

albanD commented:
> Something like add_ only changes the values of the tensors, which as far as fake tensors go is irrelevant

The original Tensor is not Meta though? And it's values won't be changed by that inplace op.

> I think this would be an unlikely case, as you're unlikely to resize_ on your module parameters or inputs.

`resize_` of the Module buffer used to be something that was commonly happening.
In any case, I think the right mindset for this kind of feature is more: the user should not be able to get the wrong result. Not that the user is unlikely to encounter wrong result.

ezyang commented:
Yeah, the specifics of the examples matter.

For the first example, it seems like views in fake mode should just be normal views. There isn't any cost to doing this for doing shape prop, since viewing is O(1).

For the second example, I don't see a way to do it without turning the assert off. 

Just a factual note:

> Isn't a simple fix to create the Tensor inside your context?

No, because the point of fake tensor mode is you can turn it on arbitrarily with any regular tensors in the environment. So no you can't move it.

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

ezyang commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -45,6 +47,11 @@ namespace{
 namespace {
+
+bool torch_dispatch_set() {
+  return static_cast<bool>(at::impl::TorchDispatchModeTLS::get_state());
```
This would be more general if we also test for Python key on the argument tensors as well, as the assert disabling here only works for modes.

albanD replied:
Just to confirm we agree on this. These assert are failing for a valid reason here right? The view semantic is not properly implemented by this mode (view of an outside Tensor will not be properly updated if an inplace happens inside). So this disable is only intended to unblock experimentation while we work on a fix for that?

albanD replied:
I definitely agree that they are very restrictive and can be annoying when trying things out. But that doesn't make them wrong. And for sure not a good reason to just disable them.
Maybe we want to add an env variable to disable them to ease local development?

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -97,11 +97,15 @@ def _enable_mode(mode, mode_info: _ModeInfo, *, replace=None, ignore_preexisting
     finally:
         mode_info.set_mode(old)
+        if hasattr(mode, "cleanup_mode"):
+            mode.cleanup_mode()  # type: ignore[attr-defined]
```
We should be able to get rid of this under the new design right

eellison replied:
yup - working on cleaning up, had to debug issue with new-style modes. when i requested another review didn't realize there was workaround to bug I hit

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

Commit:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

eellison committed:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

eellison committed:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

eellison committed:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

eellison committed:
Update on "Add Caching of Conversion to Fake/Meta tensors in FakeTensorMode"


This PR does a few things to allow caching of conversion to Fake/Meta Tensors in order for the output FakeTensors to share storage/have accurate aliasing relationships (happy to break up prs as needed but they are all pretty intertwined): 

- invokes `setup_mode` and `cleanup_mode` in the invocation of `_enable_mode`. This is needed to set up a cache of Fake/Meta Tensor conversions. 

- Disables the debug invariant checking in `VariableType` that checks things like storage/tensor ptr counts == 1, or that the input and output share the same storage when a torch dispatch mode is set. These assertions limit what you can do with `torch_dispatch_mode`, and are only being called in DEBUG builds anyway. Maybe we could also have modes opt/into out of these assertions.

- Introduces a `FakeTensorConverter` similar to [MetaTensorConverter](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L77) which caches conversions of Tensors to `FakeTensors` and uses `MetaTensorConverter` under the hood so that newly allocated FakeTensors will have the same storage. There is one active `FakeTensorConverter` for the duration of `FakeTensorMode`. Since all newly allocated tensors will be on `meta` devices, memory should not significantly increase.

The end result of the PR is so that you can do things like 
```
x = torch.rand([4, 4])

with enable_torch_dispatch_mode(FakeTensorMode):
     # conversion from x to Meta/Fake cached for duration of the `FakeTensorMode` call
     y = x[0]
     z = x[1]

self.assertEqual(torch._C._storage_id(y), torch._C._storage_id(z))
```

[ghstack-poisoned]

eellison commented:
@pytorchbot merge this please

eellison commented:
@pytorchbot merge
